proposals/0759-react-native-frameworks.md

```diff
@@ -0,0 +1,452 @@
+
+For those users we will:
+
+- Move the current template **outside** of the `react-native` package:
+  - Specifically move it to `@react-native-community/template` and kept on being **maintained by the community and the React Native team**.
+    - This is the least disruptive change for users, as it will keep on being easily accessible but it will be clearer that is a community owned component, and not part of the core of react-native
+    - Moreover, the template will be closer to where the community CLI lives today (`@react-native-community/cli`)
+  - We (Meta & the community) want to keep on **maintaining** the template and make sure that users that can’t migrate to a framework, still have a viable alternative.
+  - We don’t foresee the template to change drastically in the future. Only **minimal changes** will be allowed to the template, to reduce the upgrade pain between versions.
+  - Similarly, we’ll aim at **reducing the breaking changes** as much as possible.
+- Update all the various community tools to refer to the new template location (e.g. the Upgrade Helper)
+- Work with existing React Native Framework to identify use cases that are not properly covered by their framework
+
+#### Community CLI
+
+The React Native Community CLI is tightly coupled to the template. Apps like RN-Tester (and the proposed HelloWorld app) are using the community CLI to build and being tested on debug/release on CI. Similarly the community CLI is used to do release testing.
+
+On top of this, having the community CLI living in a separate repository (`@react-native-community/cli`) introduces a number of **circular dependencies** and friction as part of the release process.
+
+Similar to the Template, we envision having a **private CLI** inside the react-native repository to support the internal use cases mentioned above. This will be achieved by evolving each command of the CLI as follows:
+
+##### Commands to be moved inside react-native core
+
+Commands that live inside `react-native` should be extensible enough to be **used by Frameworks or libraries, **if they wish to**.**
+
+- `clean, build-android, build-ios, run-android, run-ios, log-ios, log-android`
+  - Those are wrappers around the native build system and need to be moved inside the react-native repository.
+- `start, bundle`
+  - Those are wrappers around Metro and [are effectively moved](https://github.com/facebook/react-native/pull/38795) inside the `react-native` repo already.
+- `profile-hermes`
+  - Should be moved inside react-native as part of default debugging/profiling experience with Hermes
+
+##### Commands to be removed
+
+- `upgrade`
+  - This command has been removed and now only points to the upgrade helper
+  - Potentially can be removed entirely
+- `ram-bundle`
+  - Can be removed as not relevant anymore
+
+##### Commands to be left inside @react-native-community/cli
+
+- `config` and Auto Linking of dependencies.
```

Comment:
My only worry is the potential future fragmentation in available module structures and their support by frameworks. Let's imagine we have 4 frameworks in about 2 years from now. If there's no standard way to link native dependencies, it may impose additional maintenance burden on library authors to support all popular architectures and cause confusion amongst developers when installing a library. 

Theoretically, there isn't much that can go wrong here. As long as we're using CocoaPods and Gradle under the hood, all solutions would be similar as there's a limited number of scenarios to support.

I wonder whether Community CLI and some of its internals could be rebranded as "Bare React Native Framework" package that other framework developers can use to build on top. It could provide some essential requirements as to what workflows have to be supported in order to pass the compliance test. It could also be one of the requirements when endorsing a framework that it must support certain basic workflows, such as auto-linking dependencies.

Reply:
I agree on this one.

Let's say tomorrow there's a new framework that chooses to have a custom CLI and autolinking system because they truly think it's the best way to improve their users' DX. This decision will create additional work for library maintainers, as they will have to develop new, specific plugins.

That's already (kinda) an issue with Expo, as some libraries require editing additional "native" files. Expo has its custom API for that (using `@expo/config-plugins`), which actually helps users a lot. However, it's not standardized. Imagine four or five frameworks each having their custom way to create these kinds of plugins.

Reply:
> Let's imagine we have 4 frameworks in about 2 years from now. If there's no standard way to link native dependencies, it may impose additional maintenance burden on library authors to support all popular architectures and cause confusion amongst developers when installing a library.

I think there could potentially be room for something like autolinking in core, but I'm hesitant to suggest that we should use what we have today, because it is rather slow and limited (eg: it's the main speed bottleneck for us in [@expo/fingerprint](https://github.com/expo/expo/tree/main/packages/@expo/fingerprint) and does not support transitive dependency linking). Another possible approach is that React Native doesn't end up providing an implementation of autolinking directly, but instead defines a common schema that library authors can use and frameworks can build support for.

> That's already (kinda) an issue with Expo, as some libraries require editing additional "native" files. Expo has its custom API for that (using @expo/config-plugins), which actually helps users a lot. However, it's not standardized. Imagine four or five frameworks each having their custom way to create these kinds of plugins.

If another framework is interested in following the same sort of pattern that Expo established with [CNG](https://docs.expo.dev/workflow/continuous-native-generation/) then I know that we at Expo would be happy to work with them to see where we should make parts of this toolchain more generic to minimize the maintenance costs for library authors. We want to and are incentivized to ensure that support for CNG is as easy as possible for library authors, because this provides a better experience for all users of the feature.

Additionally, it's worth noting that it's also possible to use `@expo/config-plugins` in other frameworks, see the usage in [react-native-test-app](https://github.com/microsoft/react-native-test-app) for reference, or [byCedric/custom-prebuild-example](https://github.com/byCedric/custom-prebuild-example) which demonstrates using config plugins to generate macOS and Windows projects.

@blakef summarized where we landed on this topic in a previous RFC: https://github.com/react-native-community/discussions-and-proposals/pull/637#issuecomment-1559339292

Reply:
This could be included in the `@react-native-community/cli-tools` package and benefits the whole ecosystem right now.

I already do similar stuff in the [react-native-bootsplash CLI](https://github.com/zoontek/react-native-bootsplash/blob/master/react-native.config.js), would love to have this available and white label, as it's currently kinda hard for me to write a plugin for expo without duplicating the whole logic.

Imagine being able to edit any `app.json` like this:

```json
{
  "name": "RNBootSplashExample",
  "displayName": "RNBootSplashExample",

  "plugins": [
    [
      "react-native-bootsplash",
      {
        "background": "#F5FCFF",
        "logo": "./svgs/light_logo.svg",
        "logoWidth": 100
      }
    ]
  ]
}
```

Reply:
> Another possible approach for core is that React Native doesn't end up providing an implementation of autolinking directly, but instead defines a common schema that library authors can use and frameworks can build support for.

That's the direction where we're heading towards.
Practically autolinking as it is today, it's composed by two components:

1. Integration points - They should stay inside `react-native`. Those are APIs like `ReactPackage`. Practically speaking that's where you can 'feed' modules/components/libraries to core so that they're accounted as part of the build and runtime of react native.
2. Discovery logic - This should stay outside `react-native`. This is where `react-native config` and other discovery mechanism are executed (i.e. on Android we search for `.java`/`.kt` files which implements the `ReactPackage` API exposed by core). Here frameworks can implement their own discovery mechanisms.